src/config.rs

```diff
@@ -47,6 +50,23 @@ pub struct LayeredConfigs {
     arg_overrides: Option<config::Config>,
 }
 
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum ConfigSource {
+    Default,
+    Env,
+    // TODO: Track explicit file paths when user config is a dir.
+    User,
+    Repo,
+    CommandArg,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct AnnotatedConfigEntry {
+    pub name: Option<String>,
+    pub value: config::Value,
+    pub source: ConfigSource,
+}
```

Comment:
Keeping the top-level `source` wouldn't be enough. Suppose we'll add an option to show `source`, we'll need a tree of `(value, source)`, not a tree of `value` and the top-level `source`.

```
struct Annotated { .., value, source }
enum Value {
    String(..),
    ..
    Table(HashMap<String, Annotated>),
}
```

FWIW, `config::Value` is structured this way, but its `origin: Option<String>` is private and unreliable.

Reply:
Yep right, since the final resolved config can have a single table getting keys from different sources. Fixing that part too.